h2o-algos/src/main/java/hex/glm/GLM.java

```diff
@@ -141,6 +142,87 @@ public void computeCrossValidation() {
     super.computeCrossValidation();
   }
 
+
+  /* This method aligns the submodels across CV folds.
+   *
+   * In the following depiction each submodel is represented by a tuple (alpha_value, lambda_value). Ideally each CV fold
+   * should have the same submodels but there can be differences due to max_iteration and max_runtime_secs constraint.
+   * Iterations are reset when alpha value change which can cause missing model in the middle and max_runtime_secs constraint
+   * can cause missing submodels at the end of the submodel array.
+   *
+   *  CV1:
+   *   +------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *   | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 | 1, 0.6 | 1, 0.5 | 1, 0.4 |
+   *   +------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  CV2:
+   *   +------+--------+--------+--------+--------+--------+------+--------+--------+--------+
+   *   | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 0, 0.5 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 |
+   *   +------+--------+--------+--------+--------+--------+------+--------+--------+--------+
+   *
+   *    ||
+   *    ||
+   *   \  /
+   *    \/
+   *
+   *  Aligned Submodels:
+   *
+   *  CV1:
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 |  NULL  | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 | 1, 0.6 | 1, 0.5 | 1, 0.4 |
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *
+   *  CV2:
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *  | 0, 1 | 0, 0.9 | 0, 0.8 | 0, 0.7 | 0, 0.6 | 0, 0.5 | 1, 1 | 1, 0.9 | 1, 0.8 | 1, 0.7 |  NULL  |  NULL  |  NULL  |
+   *  +------+--------+--------+--------+--------+--------+------+--------+--------+--------+--------+--------+--------+
+   *
+   */
+  private double[] alignSubModelsAcrossCVModels(ModelBuilder[] cvModelBuilders) {
+    // maximum index of alpha change across all the folds
+    int[] alphaChangePoints = new int[_parms._alpha.length + 1];
+    int[] alphaSubmodels = new int[_parms._alpha.length];
+    for (int i = 0; i < cvModelBuilders.length; ++i) {
+      GLM g = (GLM) cvModelBuilders[i];
```

Comment:
Thank you @wendycwong for the weekend review!


> So, instead of searching through all alpha and lambda combos again for CV, look at the best_submodel_idx submodel, look at the submodel, take that alpha value to use instead of having to search through all. Then collect the lambdas associated with that alpha value. Please see the pictures in my comments. If between multiple runs, you do see different alphas that is the best submodel, you then will add that alpha to the list.

That's what this part does: https://github.com/h2oai/h2o-3/pull/6102/files#diff-bfdd0a804deb47b13d4ca209feaa9c26d2845ef8192f44c23bc4d5ab6caac585R183-R207

> In addition, when you pick the lambdas after deciding on an alpha, you only need to take the lambdas that the best_submodel_idx dictates. No need to take all the lambdas in since overfitting is happening with the later submodel lambdas

Unfortunately, `No need to take all the lambdas in since overfitting is happening with the later submodel lambdas` happens to be true if and only if the lambdas are sorted but they are in a user-specified order (if user specifies them).

I think possible optimization could be looking at submodels that are between the best lambda and alpha values. Something like this:
```python
alphas = [sm.alpha for sm in best_submodels_across_CVs]
lambdas = [sm.lambda for sm in best_submodels_across_CVs]

submodels_to_explore = [sm for sm in submodels 
                        if min(alphas) <= sm.alpha <= max(alphas)
                        and min(lambdas) <= sm.lambda <= max(lambdas)] 
```

I am not 100% sure that we won't miss the optimum using this heuristic but I think it would be safer than just looking at the alpha and lambda pairs from the best submodels.

My artificial (potentially unrealistic) example: 
```
best submodels:
cv 1: alpha = 0.5; lambda = 0.5
cv 2: alpha = 0.7; lambda = 0.3
```
Can we say that the optimal values for the final model won't be alpha == 0.6, lambda == 0.4? To be clear, I am not suggesting to explore unspecified alpha/lambda combination but if we'd already created a submodel with that combination, I think it should be explored when picking the final parameters.

Reply:
In your example with two sets of values, the main model will be built with alpha=0.5, lambda=0.5 and alpha = 0.7, lambda=0.3.  My proposal is to build main model only with the best model alphas derived from all the folds.  Sometimes we may only have one alpha value, sometimes more.  It does not matter.  We will run them all.  What I don't want is to run all the alphas regardless of how they perform during the cross-validation runs.  Doing it this way, you won't miss anything.

Reply:
I think you misunderstood me @wendycwong. I think the optimization that you propose - to pick just one alpha is good but what I am suggesting is not about building the main model, it concerns only the averaging of the test deviance used to pick the alpha/lambda values for the main model.

```mermaid
flowchart TB

s1(Step 1: Gather submodels with winning alphas) ---
s2(Step 2: Align them) ---
s3(Step 3 :Look across all the CVs to pick the best submodel) -- take only one alpha value ----
s4(Step 4: Build the main model)
```

My question was about **Step 1** - maybe instead of looking only at submodels with the best alphas, I should look at the alphas that are between the minimal and maximal best alpha. This should be relatively cheap since (most of) the submodels are already calculated so it only involves computation in the **Step 3** where we average the test deviance that we get from all the CV submodels to pick the optimal submodel.

If I understand you correctly you want to do the thing that I have in the diagram between the **Step3** and **Step4**. If that's the case, I already implemented that part in a [draft PR that concerns limiting the max iteration](https://github.com/h2oai/h2o-3/pull/6069/files#diff-bfdd0a804deb47b13d4ca209feaa9c26d2845ef8192f44c23bc4d5ab6caac585R295-R302) . I can move it in this PR if you think it would be better in this PR.

Reply:
I just noticed another bug that I think should be addressed in this PR:
```java
int bestId1se = bestId;
while (bestId1se > 0 && _xval_deviances[bestId1se - 1] <= bestDev1se)
  --bestId1se;
_lambdaCVEstimate = ((GLM) cvModelBuilders[0])._model._output._submodels[bestId].lambda_value;
_bestCVSubmodel = bestId;
_model._output._lambda_1se = bestId1se; // submodel ide with bestDev+one sigma
```

The `_lambda_1se` will not be set correctly if the lambdas aren't sorted.

Do you think it would be ok to let `_model._output._lambda_1se == bestId` if the lambdas aren't sorted (that would be very easy fix)? Or I can do it properly - get indices of descending lambdas and go through that. But there would be new corner cases e.g. some lambdas could be not evaluated in this way: `lambdas = [0, 1, 0.5, 0.1, 0.2, 0.3, early stop, 0.4]` etc. so we wouldn't be able to guarantee anyway that there isn't a better `lambda_1se`.

Reply:
@tomasfryda I like the diagram, it makes our decision much easier.  Yes, there is misunderstanding between us regarding what is going on.  What I mean to say is the following

We have 5 alphas in the array with lambda_search on: 

Run cross-validation and it generates cv_model_1, cv_model_2, cv_model_3.

Say cv_model_1 pick alpha0 and a lambda as best submodel;
cv_model_2 pick alpha1 and a lambda as best submodel;
cv_model_3 pick alpha0 and a lambda as best submodel;

From your step 1, step 2 and step 3, we will gather alpha0 with its lambdas to run, alpha1 with its lambdas to run.

The misunderstanding here in building the main model:
In step 4: the main model will be built with alpha0 with its lambdas picked during model alignment, alpha1 with its lambdas picked during model alignment.

Doing it this way, we are not skipping any runs with good results and hence we will not be missing anything.

What do you think?  Let me know.  Wendy

Reply:
To me, it seems that using more alphas and lambdas defeats the purpose of cross-validation. If I didn't know the GLM implementation I'd guess that in **Step 3** we pick from all the evaluated submodels across all the folds just one pair of alpha and lambda values. (Using CV to pick the best hyper-parameters)  After all the **Step 3** picks just one best submodel (`bestId`, `_selected_submodel_idx`).

~It's unclear to me how to pick the better alpha/lambda value when training the main model. Wouldn't that lead to overfitting? Maybe we use some information criterion (AIC/BIC/WAIC) to select the final main submodel but I didn't notice anything like that in the code.~ (EDIT: looks like we pick the best alpha and lambda pair in the `cv_computeAndSetOptimalParameters` and then we train main model with multiple submodels even though  we won't use them (see below))

The more I think about it the more I think we should pick just one alpha and one lambda value from the  **Step 3** and use that as the main model.


So my proposal would be:

1. (Optional) Pick alphas depending on the best submodels across the folds, this could be either just the alphas associated with the best submodels or a range between minimal alpha from the best submodels and maximal alpha from the best submodels. This would decrease the amount of the submodels to be computed in the cv_computeAndSetOptimalParameters (`g._driver.computeSubmodel`).
2. Look at all available (or selected from previous step) submodels and calculate the `_xval_deviances` and `_xval_sd`.
3. Pick the best submodel using the `_xval_deviances`
4. Set the alpha and lambda parameter for the main model obtained from the best submodel

Which shouldn't change the way it works in general actually as we use `_bestCVSubmodel` to pick the submodel for the main model:
https://github.com/h2oai/h2o-3/blob/088927f1fc86b8fe49763dfb7eda775a13a75993/h2o-algos/src/main/java/hex/glm/GLM.java#L2932-L2935

There are however some minor annoyances to solve if we decide to go this way - things that depend on submodels, e.g., `lambda_1se`.
```java
    public double lambda_1se(){return (_lambda_1se==-1 || _submodels.length==0 || _lambda_1se>=_submodels.length) ?
            -1 : _submodels[_lambda_1se].lambda_value;}
```
This ^^^ would be a problem since we wouldn't train submodels `0..._best_submodel_idx-1` so we could not look up the lambda value in submodels as there would be just one submodel trained (the best one picked in `cv_computeAndSetOptimalParametes`.)

I am currently looking more into the possibility of implementing "my" proposal (how many things would need to change, if it really does what GLM should do just with less computation etc.).